pkg/oomstore/cache.go

```diff
@@ -0,0 +1,175 @@
+package oomstore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/oom-ai/oomstore/internal/database/dbutil"
+	"github.com/oom-ai/oomstore/internal/database/metadata"
+	"github.com/oom-ai/oomstore/internal/database/offline"
+	"github.com/oom-ai/oomstore/pkg/errdefs"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types"
+)
+
+const (
+	Day       = 24 * time.Hour
+	Capacity  = 1000
+	Period    = 5 * time.Minute
+	MinPeriod = 2 * time.Minute
+)
+
+type BufferGroup struct {
+	groupID  int
+	records  []types.StreamRecord
+	lastPush time.Time
+	mu       sync.Mutex
+}
+
+type Cache struct {
+	capacity  int
+	minPeriod time.Duration
+	mu        sync.Mutex
+	quit      chan bool
+
+	ticker *time.Ticker
+	ch     chan types.StreamRecord // buffer channel
+	buffer map[int]*BufferGroup    // group_id -> buffered rows
+}
+
+func (s *OomStore) InitCache(ctx context.Context) {
+	cache := &Cache{
+		// TODO: make Capacity, Period, MinPeriod configurable
+		capacity:  Capacity,
+		minPeriod: MinPeriod,
+		quit:      make(chan bool),
+
+		ch:     make(chan types.StreamRecord),
+		buffer: make(map[int]*BufferGroup),
+		ticker: time.NewTicker(Period),
+	}
+	s.cache = cache
+
+	go func() {
+		for {
+			select {
+			case <-cache.quit:
+				return
+			case <-cache.ticker.C:
+				for groupID, b := range cache.buffer {
+					if len(b.records) > 0 && time.Since(b.lastPush) > cache.minPeriod {
+						if err := cache.send(ctx, s, groupID); err != nil {
+							fmt.Printf("%+v", err)
+						}
+					}
+				}
+			case record := <-cache.ch:
+				groupID := record.GroupID
+				if len(cache.buffer[groupID].records) == cache.capacity {
+					if err := cache.send(ctx, s, groupID); err != nil {
+						fmt.Printf("%+v", err)
+					}
+				}
+				cache.buffer[groupID].records = append(cache.buffer[groupID].records, record)
+			}
+		}
+	}()
+}
+
+func (c *Cache) Close() error {
+	c.ticker.Stop()
+	c.quit <- true
+	return nil
+}
+
+func (c *Cache) Push(groupID int, record types.StreamRecord) {
```

Comment:
`types.StreamRecord` already has the field `GroupID`, so you can remove the `groupID` parameter?

Reply:
fixed